### PR TITLE
update go to 1.18

### PR DESCRIPTION
**Pull request by lennysgarage**
Signed-off-by: Jonathan Marcantonio <jmarcant@redhat.com>

https://github.com/stolostron/backlog/issues/23955

### CODE REVIEW

File: pkg/controller/mcmhub/hook_test.go [hunk 1]
```diff
@@ -262,14 +261,14 @@ var _ = Describe("given a subscription pointing to a git path without hook folde
...
-				return fmt.Errorf("failed to get subscription %s, err: %s", subKey, err.Error())
+				return fmt.Errorf("failed to get subscription %s, err: %w", subKey, err)
 			}
 
 			fmt.Printf("subscription= %+v\n", u)
 
 			appsubReport := &appsubReportV1alpha1.SubscriptionReport{}
 			if err := k8sClt.Get(ctx, appsubReportKey, appsubReport); err != nil {
-				return fmt.Errorf("failed to get the appsub AppsubReport %s, err: %s", appsubReportKey, err.Error())
+				return fmt.Errorf("failed to get the appsub AppsubReport %s, err: %w", appsubReportKey, err)
```
Comment (mikeshng) [inline pkg/controller/mcmhub/hook_test.go:271]: Why is this change needed? It seems fine before?

File: pkg/controller/mcmhub/hook_test.go [hunk 2]
```diff
@@ -262,14 +261,14 @@ var _ = Describe("given a subscription pointing to a git path without hook folde
 			u := &subv1.Subscription{}
 
 			if err := k8sClt.Get(ctx, subKey, u); err != nil {
-				return fmt.Errorf("failed to get subscription %s, err: %s", subKey, err.Error())
+				return fmt.Errorf("failed to get subscription %s, err: %w", subKey, err)
```
Comment (mikeshng) [inline pkg/controller/mcmhub/hook_test.go:264]: Why is this change needed? It seems fine before?

File: pkg/controller/mcmhub/hub.go [hunk 3]
```diff
@@ -75,7 +71,7 @@ func (r *ReconcileSubscription) doMCMHubReconcile(sub *appv1alpha1.Subscription)
 
 	chnAnnotations := primaryChannel.GetAnnotations()
 
-	if chnAnnotations[appv1.AnnotationResourceReconcileLevel] != "" {
+	if chnAnnotations[appv1alpha1.AnnotationResourceReconcileLevel] != "" {
```
Comment (mikeshng) [inline pkg/controller/mcmhub/hub.go:74]: We should keep `appv1 "open-cluster-management.io/multicloud-operators-subscription/pkg/apis/apps/v1"` and delete the `appv1alpha1 "open-cluster-management.io/multicloud-operators-subscription/pkg/apis/apps/v1"`

Because appv1 is the current version. The alpha is the old version that we no longer use.

File: pkg/controller/mcmhub/hub.go [hunk 4]
```diff
@@ -196,7 +192,7 @@ func (r *ReconcileSubscription) GetChannelNamespaceType(s *appv1alpha1.Subscript
 	return chNameSpace, chName, chType
 }
 
-func GetSubscriptionRefChannel(clt client.Client, s *appv1.Subscription) (*chnv1.Channel, *chnv1.Channel, error) {
+func GetSubscriptionRefChannel(clt client.Client, s *appv1alpha1.Subscription) (*chnv1alpha1.Channel, *chnv1alpha1.Channel, error) {
```
Comment (mikeshng) [inline pkg/controller/mcmhub/hub.go:195]: same comment as above we should keep the chnv1 and remove the alpha1 api

File: pkg/subscriber/git/git_subscriber_test.go [hunk 5]
```diff
@@ -536,8 +534,8 @@ var _ = Describe("github subscriber reconcile options", func() {
 		"propagate them to the synchronizer", func() {
 
 		subAnnotations := make(map[string]string)
-		subAnnotations[appv1.AnnotationClusterAdmin] = "true"
-		subAnnotations[appv1.AnnotationResourceReconcileOption] = "merge"
+		subAnnotations[appv1alpha1.AnnotationClusterAdmin] = "true"
```
Comment (mikeshng) [inline pkg/subscriber/git/git_subscriber_test.go:537]: same as above comment keep the appv1 and delete the alpha1 api

File: pkg/utils/gitrepo.go [hunk 6]
```diff
@@ -827,7 +827,7 @@ func sortKubeResource(crdsAndNamespaceFiles, rbacFiles, otherFiles []string, pat
 			if err != nil {
 				klog.Warning("Failed to unmarshal YAML file")
 				// Just ignore the YAML
-				return crdsAndNamespaceFiles, rbacFiles, otherFiles, nil
+				return crdsAndNamespaceFiles, rbacFiles, otherFiles, err
```
Comment (mikeshng) [inline pkg/utils/gitrepo.go:830]: This might not be correct since there is a warning up front and and it might be intentional that we are not returning the err here.

If you want to use the err then put it in the warning message is suffice. 
If you think it should return with err, maybe a separate PR is better.

**Review by mikeshng — COMMENTED** [review]
This is really good work. Just some comments.

/hold

File: pkg/controller/mcmhub/hub.go (hunk 3 above)
Comment (lennysgarage) [reply pkg/controller/mcmhub/hub.go:74]: Thank you will change this

File: pkg/controller/mcmhub/hook_test.go (hunk 1 above)
Comment (lennysgarage) [reply pkg/controller/mcmhub/hook_test.go:271]: This change came from adding the linter, `errorlint` which recommended a non-wrapping format as well when researching why it wanted %w I came along this: https://stackoverflow.com/a/61287626

File: pkg/placementrule/utils/auth.go [hunk 7]
```diff
@@ -44,7 +44,7 @@ var (
 func FilteClustersByIdentity(authClient kubernetes.Interface, object runtime.Object, clmap map[string]*spokeClusterV1.ManagedCluster) error {
 	objmeta, err := meta.Accessor(object)
 	if err != nil {
-		return nil
+		return err
```
[the remark this answers is not in the document]
Comment (lennysgarage) [reply pkg/placementrule/utils/auth.go:47]: Reverting change.

File: pkg/utils/reference.go [hunk 8]
```diff
@@ -212,7 +212,7 @@ func DeleteReferredObjects(clt client.Client, rq types.NamespacedName, gvk schem
 			u.SetOwnerReferences(newOwers)
 			err := clt.Update(context.TODO(), u)
 			if err != nil {
-				return nil
+				return err
```
[the remark this answers is not in the document]
Comment (lennysgarage) [reply pkg/utils/reference.go:215]: Reverting change.

**Review by mikeshng — APPROVED** [review]
/approve

/lgtm

thanks!
